File: 0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string_view s) {
        const int n = static_cast<int>(s.size());
        int best = 0;

        for (int i = 0; i < n; ++i) {
            unordered_set<char> seen;
            // Early cut‑off: if the remaining chars can’t beat current best
            if (n - i <= best) break;

            for (int j = i; j < n; ++j) {
                char c = s[j];
                if (seen.count(c)) break;      // duplicate found, stop growing
                seen.insert(c);
                best = max(best, j - i + 1);   // update answer
            }
        }
        return best;
    }
};

File: 0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        cnt={}
        l=0
        n=len(s)
        ans=0
        for i,ch in enumerate(s):
            if(ch in cnt):
                cnt[ch]+=1
            else:
                cnt[ch]=1
            while(cnt[ch]>1):
                cnt[s[l]]-=1
                l+=1
            ans=max(ans,i-l+1)
        return ans    
